mitodl/ocw-studio
# end to end testing pipeline

# What are the relevant tickets?
Closes https://github.com/mitodl/ocw-studio/issues/2033

# Description (What does it do?)
This PR adds a new pipeline definition, `TestPipelineDefiniton`. This pipeline is designed for running the end to end Playwright tests defined in `ocw-hugo-themes` against the output of the steps defined in `SitePipelineOnlineTasks`. The resulting pipeline will be automatically triggered on pushes to `ocw-hugo-themes` or `ocw-hugo-projects` based on configured branches and will alert us of any problems running the end to end tests against content that was built and deployed using the same steps we use to build and deploy any other site using pipelines generated by `ocw-studio`.

# How can this be tested?
 - Make sure your instance of `ocw-studio` is set up to publish websites according to the readme
 - Set the following environment variables:
```
OCW_WWW_TEST_SLUG=ocw-ci-test-www
OCW_COURSE_TEST_SLUG=ocw-ci-test-course
AWS_TEST_BUCKET_NAME=ocw-content-test
AWS_OFFLINE_TEST_BUCKET_NAME=ocw-content-offline-test
```
 - Download the JSON fixtures from this Google Drive folder: https://drive.google.com/drive/folders/1iqLw2C2Atk1_O94z7PpUC9X7JQe_IxSn
 - In your local Django admin, make sure you don't have existing sites that would conflict with the test courses. You can do this by going to http://localhost:8043/admin/websites/website/ and looking for `ocw-ci-test-course` and `ocw-ci-test-www`. If these websites already exist, delete them.
 - Place `test_websites.json` and `test_website_content.json` at the root of your `ocw-studio` folder
 - Run `docker compose exec web ./manage.py loaddata test_websites.json` and you should see `Installed 2 object(s) from 1 fixture(s)`
 - Run `docker compose exec web ./manage.py loaddata test_website_content.json` and you should see `Installed 43 object(s) from 1 fixture(s)`
 - Run `docker compose exec web ./manage.py reset_sync_states --filter "ocw-ci-test-course, ocw-ci-test-www" --skip_sync`
 - Run `docker compose exec web ./manage.py sync_website_to_backend --filter "ocw-ci-test-course, ocw-ci-test-www" --delete`
 - Browse to http://localhost:8043/sites/ocw-ci-test-www and publish the site live. Don't worry if the publish fails, the important part is that it gets pushed to the `release` branch in Github. To verify this browse to the repo in your test org and make sure that the site content is in the `release` branch.
 - Repeat the previous step for http://localhost:8043/sites/ocw-ci-test-course
 - Download the test course static resources from this Google Drive link: https://drive.google.com/drive/folders/14Hlid31Qy7Yy5V4OgHUwNleYUFuJ5BH2?usp=sharing
 - Open the Minio web UI at http://localhost:9001 and login with your credentials
 - Browse to the `ol-ocw-studio-app` bucket, then in the `courses` folder create a folder named `ocw-ci-test-course` if it doesn't already exist
 - Drag and drop the static resources we downloaded from Google Drive into the Minio folder we just created
 - Run `docker compose exec web ./manage.py upsert_theme_assets_pipeline --themes-branch cg/playwright-external-site`
 - Run `docker compose exec web ./manage.py upsert_e2e_test_pipeline --themes-branch cg/playwright-external-site`
 - Browse to your locally running Concourse instance at http://localhost:8080
 - Find the theme assets pipeline that we pushed up for `cg/playwright-external-site`, unpause it and trigger a build. The theme assets build has been altered as part of this PR to also copy the theme assets to the test buckets. Wait for it to finish.
 - Once the theme assets build has completed, locate the end to end testing pipeline and unpause it. The pipeline should eventually start automatically after about a minute. If it does not, check the inputs to the job and try refreshing them manually. Once the pipeline starts, it should complete without error.

# Additional Context
The offline build is currently not tested, but will be tested in the future and this pipeline will be updated to support it. Before this PR is merged, the new settings should be set in Heroku via a `salt-ops` PR. These settings include:

 - `OCW_WWW_TEST_SLUG`
 - `OCW_COURSE_TEST_SLUG`
 - `AWS_TEST_BUCKET_NAME`
 - `AWS_OFFLINE_TEST_BUCKET_NAME`

This `salt-ops` PR will not be created until https://github.com/mitodl/ol-infrastructure/issues/1945 is resolved and the new buckets that it will point at have been created.
